odl/trafos/backends/pyfftw_bindings.py

```diff
@@ -0,0 +1,309 @@
+            axes = tuple(axes)
+
+    direction = _pyfftw_to_local(direction)
+    fftw_plan_in = kwargs.pop('fftw_plan', None)
+    planning_effort = _pyfftw_to_local(kwargs.pop('planning_effort',
+                                                  'estimate'))
+    planning_timelimit = kwargs.pop('planning_timelimit', None)
+    threads = kwargs.pop('threads', None)
+    normalise_idft = kwargs.pop('normalise_idft', False)
+    wimport = kwargs.pop('import_wisdom', '')
+    wexport = kwargs.pop('export_wisdom', '')
+
+    # Cast input to complex if necessary
+    array_in_copied = False
+    if is_real_dtype(array_in.dtype) and not halfcomplex:
+        # Need to cast array_in to complex dtype
+        array_in = array_in.astype(TYPE_MAP_R2C[array_in.dtype])
+        array_in_copied = True
+
+    # Do consistency checks on the arguments
+    _pyfftw_check_args(array_in, array_out, axes, halfcomplex, direction)
+
+    # Import wisdom if possible
+    if wimport:
+        try:
+            with open(wimport, 'rb') as wfile:
+                wisdom = pickle.load(wfile)
+        except IOError:
+            wisdom = []
+        except TypeError:  # Got file handle
+            wisdom = pickle.load(wimport)
+
+        if wisdom:
+            pyfftw.import_wisdom(wisdom)
+
+    # Copy input array if it hasn't been done yet and the planner is likely
+    # to destroy it. If we already have a plan, we don't have to worry.
+    planner_destroys = _pyfftw_destroys_input(
+        [planning_effort], direction, halfcomplex, array_in.ndim)
+    must_copy_array_in = fftw_plan_in is None and planner_destroys
+
+    if must_copy_array_in and not array_in_copied:
+        plan_arr_in = np.empty_like(array_in)
+        flags = [_local_to_pyfftw(planning_effort), 'FFTW_DESTROY_INPUT']
+    else:
+        plan_arr_in = array_in
+        flags = [_local_to_pyfftw(planning_effort)]
+
+    if fftw_plan_in is None:
+        if threads is None:
+            if plan_arr_in.size < 1000:  # Somewhat arbitrary
+                threads = 1
+            else:
+                threads = cpu_count()
+
+        fftw_plan = pyfftw.FFTW(
+            plan_arr_in, array_out, direction=_local_to_pyfftw(direction),
+            flags=flags, planning_timelimit=planning_timelimit,
+            threads=threads, axes=axes)
+    else:
+        fftw_plan = fftw_plan_in
+
+    fftw_plan(array_in, array_out, normalise_idft=normalise_idft)
+
+    if wexport:
+        try:
+            with open(wexport, 'ab') as wfile:
+                pickle.dump(pyfftw.export_wisdom(), wfile)
+        except TypeError:  # Got file handle
+            pickle.dump(pyfftw.export_wisdom(), wexport)
+
+    return fftw_plan
+
+
+def _pyfftw_to_local(flag):
+    return flag.lstrip('FFTW_').lower()
+
+
+def _local_to_pyfftw(flag):
+    return 'FFTW_' + flag.upper()
+
+
+def _pyfftw_destroys_input(flags, direction, halfcomplex, ndim):
+    """Return ``True`` if FFTW destroys an input array, ``False`` otherwise."""
```

Comment:
Destroy isnt really well defined. Is the pointer freed, or is the data scrambled?